airtable/plugin.py

```diff
@@ -1,51 +1,65 @@
 import json
+import re
 from dataclasses import dataclass
 from datetime import datetime
-from typing import Any, Tuple
-import re
+from typing import Any, Optional, Tuple
+from urllib.parse import unquote, urlparse
 
 import httpx
-
 from lutraai.augmented_request_client import AugmentedTransport
 
+
 @dataclass
 class AirtableBaseID:
     id: str
 
+
 @dataclass
 class AirtableTableID:
     """
     id can either be the table name or the table ID.
     """
+
     id: str
 
+
 @dataclass
 class AirtableRecordID:
     id: str
 
 
-def airtable_url_to_ids(url: str) -> Tuple[AirtableBaseID, AirtableTableID]:
+def airtable_parse_ids_from_url(
+    url: str,
+) -> Tuple[AirtableBaseID, Optional[AirtableTableID], Optional[AirtableRecordID]]:
     """
-    Extracts the base ID and table ID from an Airtable URL.
+    Parse Airtable IDs from an Airtable web UI URL.  url must be in the format
+    "https://airtable.com/{base_id}/{table_id}/{view_id}/{record_id}" and may include
+    additional path segments or query arguments after the record ID which will be
+    ignored.
+
+    Only the base_id is required and guaranteed to be non-None.  For each other ID, if
+    it is not present in the URL, the corresponding value in the tuple will be None.
+    You must check that these ID values are not None before using them.
     """
-    # Regular expression to extract base and table IDs
-    # This pattern is flexible to accommodate prefixes before 'airtable.com'
-    pattern = r"airtable\.com/([^/?]+)(?:/([^/?]+))?"
-    match = re.search(pattern, url)
-
-    if not match:
-        raise ValueError("Invalid Airtable URL")
-
-    base_id = match.group(1)
-    # When loading the Airtable using the URL that only has the base id (e.g. https://airtable.com/apptpbyiHjjfPecFw)
-    # Airtable automatically drops you into the first table of the base and we don't expect the table_id to be an empty string
-    # if the user copies the URL from the browser.
-    if not match.group(2):
-        raise ValueError("Invalid Airtable table ID")
-
-    table_id = match.group(2)
+    parsed_url = urlparse(url)
+    if parsed_url.netloc != "airtable.com":
+        raise ValueError(f"host must be airtable.com: {url}")
+    match = re.search(
+        r"/(?P<base_id>app[\w\d]+)(?:/(?P<table_id>tbl[\w\d]+))?(?:/(?P<view_id>viw[\w\d]+))?(?:/(?P<record_id>rec[\w\d]+))?(?:/.*)?",
+        unquote(parsed_url.path),
+    )
+    if match:
+        base_id = match.group("base_id")
+        table_id = match.group("table_id")
+        record_id = match.group("record_id")
+        return (
+            AirtableBaseID(base_id),
+            AirtableTableID(table_id) if table_id else None,
+            AirtableRecordID(record_id) if record_id else None,
+        )
+    else:
+        raise ValueError(f"does not match the expected Airtable URL format: {url}")
```

Comment:
let's repeat the format here in the error

```url must be in the format
    "https://airtable.com/{base_id}/{table_id}/{view_id}/{record_id}"```

Reply:
Done.